contracts/contracts/strategies/GeneralizedUniswapV3Strategy.sol

```diff
@@ -0,0 +1,850 @@
+    }
+
+    /**
+     * @notice Change the reserve strategies of the supported assets
+     * @dev Will throw if the strategies don't support the assets
+     * @param _token0ReserveStrategy The new reserve strategy for token0
+     * @param _token1ReserveStrategy The new reserve strategy for token1
+     */
+    function _setReserveStrategy(
+        address _token0ReserveStrategy,
+        address _token1ReserveStrategy
+    ) internal {
+        require(
+            IStrategy(_token0ReserveStrategy).supportsAsset(token0),
+            "Invalid Reserve Strategy"
+        );
+        require(
+            IStrategy(_token1ReserveStrategy).supportsAsset(token1),
+            "Invalid Reserve Strategy"
+        );
+
+        reserveStrategy[token0] = _token0ReserveStrategy;
+        reserveStrategy[token1] = _token1ReserveStrategy;
+
+        emit ReserveStrategiesChanged(
+            _token0ReserveStrategy,
+            _token1ReserveStrategy
+        );
+    }
+
+    /***************************************
+            Deposit/Withdraw
+    ****************************************/
+
+    /// @inheritdoc InitializableAbstractStrategy
+    function deposit(address _asset, uint256 _amount)
+        external
+        override
+        onlyVault
+        nonReentrant
+    {
+        require(_asset == token0 || _asset == token1, "Unsupported asset");
+        IVault(vaultAddress).depositForUniswapV3(_asset, _amount);
+        // Not emitting Deposit event since the Reserve strategy would do so
+    }
+
+    /// @inheritdoc InitializableAbstractStrategy
+    function depositAll() external override onlyVault nonReentrant {
+        _depositAll();
+    }
+
+    /**
+     * @notice Deposits all undeployed balances of the contract to the reserve strategies
+     */
+    function _depositAll() internal {
+        uint256 token0Bal = IERC20(token0).balanceOf(address(this));
+        uint256 token1Bal = IERC20(token1).balanceOf(address(this));
+        if (token0Bal > 0) {
+            IVault(vaultAddress).depositForUniswapV3(token0, token0Bal);
+        }
+        if (token1Bal > 0) {
+            IVault(vaultAddress).depositForUniswapV3(token1, token1Bal);
+        }
+
+        // Not emitting Deposit events since the Reserve strategies would do so
+    }
+
+    /**
+     * @notice Withdraws asset from the reserve strategy
+     * @inheritdoc InitializableAbstractStrategy
+     */
+    function withdraw(
+        address recipient,
+        address _asset,
+        uint256 amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == token0 || _asset == token1, "Unsupported asset");
+
+        IERC20 asset = IERC20(_asset);
+        uint256 selfBalance = asset.balanceOf(address(this));
+
+        if (selfBalance < amount) {
+            // Try to pull remaining amount from reserve strategy
+            // This might throw if there isn't enough in reserve strategy as well
+            IVault(vaultAddress).withdrawForUniswapV3(
```

Comment:
These strategies are already under the vault's control, and could be withdrawn directly from. The part that the vault can't get to is the uniswap position. We'd need to withdraw from the currently active position instead. This might give us some extra coins on the other side, but that would be okay. They can sit on the strategy, because the vault might be asking for them next.

Reply:
Got it, will change that part of the code